module/surface_sam_sid_vhf.c

```diff
@@ -0,0 +1,411 @@
+/*
+ * Virtual HID Framwork (VHF) driver for input events via SAM.
+ * Used for keyboard input events on the Surface Laptops.
+ */
+
+#include <linux/acpi.h>
+#include <linux/hid.h>
+#include <linux/input.h>
+#include <linux/platform_device.h>
+#include <linux/types.h>
+
+#include "surface_sam_ssh.h"
+
+#define SID_VHF_INPUT_NAME		"Microsoft Virtual HID Framework Device"
+
+/*
+ * Request ID for VHF events. This value is based on the output of the Surface
+ * EC and should not be changed.
+ */
+#define SAM_EVENT_SID_VHF_RQID		0x0015
+#define SAM_EVENT_SID_VHF_TC		0x15
+
+struct sid_vhf_evtctx {
+	struct device     *dev;
+	struct hid_device *hid;
+};
+
+struct sid_vhf_drvdata {
+	struct sid_vhf_evtctx event_ctx;
+};
+
+
+static int sid_vhf_hid_start(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+	return 0;
+}
+
+static void sid_vhf_hid_stop(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+}
+
+static int sid_vhf_hid_open(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+	return 0;
+}
+
+static void sid_vhf_hid_close(struct hid_device *hid)
+{
+	hid_dbg(hid, "%s\n", __func__);
+}
+
+struct surface_sam_sid_vhf_meta_rqst {
+	u8 id;
+	u32 offset;
+	u32 limit;
```

Comment:
`limit` sound's a bit like a hard buffer limit. While this is true when sending the request, I think `length` might be a bit clearer that the response changes this field. Also a comment indicating that would probably be nice.